standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require uses to put down a deposit.
+The aim of the deposit initially is purely to protect the system from abuse.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+The RLN contract provides the following functionality:
+- register a membership in a given tier;
+- check if a membership is valid;
+- withdraw the deposit for an expired membership;
+- extend an expired membership.
+
+For the `Owner`, the contract provides the following additional functionality:
+- change the modifiable parameters (see parameter table below);
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing;
+- membership expiration includes a grace period;
+- a user can extend their membership without putting up any additional deposit.
+
+### Membership life-cycle
+
+To register a membership, a user puts locks up a deposit.
+Each membership has an expiration date.
+After a membership expires, it enters a grace period, during which the user can extend it.
+Extending a membership requires no additional deposit.
+After the grace period, an expired membership can be taken over by another user.
+At any point, a user can withdraw their deposit and terminate the membership.
+
+In more detail, the membership life-cycle is as follows:
+1. A user registers a membership:
+    1. If there are expired unclaimed slots in the tree, use the slot with the earliest expiration date past its grace period, otherwise create a new slot;
+    2. if the membership re-uses a previously used slot, refund the previous user’s deposit.
+2. The user sends messages while the membership is active:
+    1. If the user exceeds the rate limit, their over-limit messages are not propagated in that epoch; the deposit is not slashed.
+3. After the membership expires, the user can do one of the following:
+    1. Do nothing. The deposit remains in the contract, and the user retains the ability to send messages (respecting the per-epoch rate limit);
+    2. Withdraw the deposit. The user receives the full deposit back, and loses the ability to send messages;
+    3. Extend the membership. The user sends a transaction to re-enable their membership for another expiration term. The user only covers gas costs and does not have to provide additional funds. The original deposit remains in the contract. The membership is prolonged for another expiration term under the same conditions.
+
+The user can extend the membership both during and after its grace period.
+By not extending the membership within its grace period, the user assumes the risk of the membership being taken over at any moment.
+
+A user can hold multiple memberships.
```

Comment:
I've realized that I don't understand the relationship between a node (i.e., a node ID?) and a (blockchain) address as identifiers. What is the primary identifier here? From the blockchain standpoint, I would assume, blockchain address is what matters... Where does the node ID come into pucture?

Reply:
There are acutally two questions inside "Can a user hold multiple memberships?".

First: can the same Waku node (i.e., the same P2P node ID) hold multiple RLN memberships? The answer is: technically it could, but it's not currently supported by any Waku implementation. I don't think there's any disagreement here.

Second: if one sends two transactions that call the "register a new membership" function of the RLN contract from the same Ethereum address, should both transactions succeed? Should the contract check whether this Ethereum address has already been issued a membership?

I would suggest that the contract doesn't care which Ethereum addresses the transactions are coming from as long as they pay sufficient fees. Plase let me know if there are reasons why this is a bad idea.

Reply:
I agree with your interpretation @s-tikhomirov:

- A Waku node can only hold when RLN membership, this is an implementation limitation. This limitation might be lifted by the Waku team in the future, or anyone developer that decides to implement it themselves.

- A given Ethereum address can register several membership with the RLN smart contract: I suggest to keep this possible. I don't see a reason why we would stop it, on the contrary, it makes it easier for fleet operators to manage their nodes.

Reply:
> A given Ethereum address can register several membership with the RLN smart contract: I suggest to keep this possible.

> I think the above point would make the implementation more difficult

?

We are talking about enabling Ethereum account XYZ to register as many RLN membership as they want by sending fees + commitment to the contract.
This is already implemented.
This is the easy version.

The difficult version is updating the smart contract to keep track of the caller and limiting an insert to one membership per Ethereum address.

 > I might be missing something but in which sense the fleet management would be easier :)?

Because as a fleet owner, you can have one Ethereum wallet and use the same wallet to buy the memberships of all your nodes.
If we were to limit to one RLN account per Ethereum wallet, then as a fleet owner, you'd nee to generate and fund one ethereum wallet per RLN membership, which is cumbersome (and pointless).